src/devices/SoftwareSpi/SoftwareSpi.cs

```diff
@@ -0,0 +1,199 @@
+                    });
+            }
+            else
+            {
+                _bitTransfer = new ScopeData(
+                    exit: () => {
+                        _controller.Write(_clk, !idle);
+                        _controller.Write(_clk, idle);
+                    });
+            }
+
+            _chipSelect = new ScopeData(
+                enter: () => {
+                    _controller.Write(_cs, _settings.ChipSelectLineActiveState);
+                },
+                exit: () => {
+                    _controller.Write(_cs, !(bool)_settings.ChipSelectLineActiveState);
+                });
+        }
+
+        public override SpiConnectionSettings ConnectionSettings => _settings;
+
+        public override void TransferFullDuplex(ReadOnlySpan<byte> dataToWrite, Span<byte> dataToRead)
+        {
+            if (dataToRead.Length != dataToRead.Length)
+            {
+                throw new ArgumentException(nameof(dataToWrite));
+            }
+
+            int bitLen = _settings.DataBitLength;
+            int lastBit = bitLen - 1;
+
+            using (ChipSelect)
+            {
+                for (int i = 0; i < dataToRead.Length; i++)
+                {
+                    byte readByte = 0;
+                    for (int j = 0; j < bitLen; j++)
+                    {
+                        using (BitTransfer)
+                        {
+                            int bit = _settings.DataFlow == DataFlow.MsbFirst ? lastBit - j : j;
+                            bool bitToWrite = ((dataToWrite[i] >> bit) & 1) == 1;
+                            if (ReadWriteBit(bitToWrite))
+                            {
+                                readByte |= (byte)(1 << bit);
+                            }
+                        }
+                    }
+
+                    dataToRead[i] = readByte;
+                }
+            }
+        }
+
+        private bool ReadWriteBit(bool bitToWrite)
+        {
+            _controller.Write(_mosi, bitToWrite);
+            return (bool)_controller.Read(_miso);
+        }
+
+        public override void Read(Span<byte> data)
+        {
+            Span<byte> dataToWrite = stackalloc byte[data.Length];
+            dataToWrite.Fill(0);
+            TransferFullDuplex(dataToWrite, data);
+        }
+
+        public override void Write(ReadOnlySpan<byte> data)
+        {
+            Span<byte> dataToRead = stackalloc byte[data.Length];
+            TransferFullDuplex(data, dataToRead);
+        }
+
+        public override void WriteByte(byte data)
+        {
+            Span<byte> outData = stackalloc byte[1];
+            outData[0] = data;
+            Write(outData);
+        }
+
+        public override byte ReadByte()
+        {
+            Span<byte> data = stackalloc byte[1];
+            Read(data);
+            return data[0];
+        }
+
+        public new void Dispose()
+        {
+            _controller?.Dispose();
```

Comment:
doesn't SpiDevice implement the dispose pattern? If it doesn't I think it should

Reply:
well it's not virtual so you cannot override. I got mixed feeling if it should. I will check what we do elsewhere and follow that pattern

Reply:
Ideally we would want SpiDevice to be like:

```c#
public class SpiDevice : IDisposable
{
    public void Dispose()
    {
        Dispose(true);
        GC.SuppressFinalize(this);
    }

    protected virtual void Dispose(bool disposing)
    {
        //Dispose stuff that is specific to SpiDevice and let subclasses dispose their own stuff.
    }
}
```

Reply:
There still seems to be a problem here since this Dispose shouldn't be public, but I assume that the problem lies on the fact that SpiDevice doesn't implement the dispose pattern correctly.

Reply:
> SpiDevice doesn't implement the dispose pattern correctly.

We recently updated this.  Did I do something incorrectly here... https://github.com/dotnet/iot/pull/614/commits/6d4b8718b0da000b92156c18e9519372ff4a6f1e

Reply:
mm might be that we just haven't updated the version we ingest here of System.Device.Gpio. @krwq might be worth it to just wait and rebase tomorrow once we have ingested the new changes so this one is a protected override instead of a public override.